subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/ModuleSelectors.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder;
+
+import com.google.common.collect.Lists;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.selectors.ResolvableSelectorState;
+
+import java.util.Iterator;
+import java.util.List;
+
+public class ModuleSelectors<T extends ResolvableSelectorState> implements Iterable<T> {
```

Comment:
:o: I think unit tests for the public methods of this class (custom `add`, `remove`) would be nice to have. Since these are performance optimized implementations of otherwise trivial functionality. If something is changed here in the future, I am sure one of the other tests will fail, but it might be hard to track down the problem then.
The unit tests could cover the different branches (no selector, one dynamic selector, one normal selector, more than one selector, ...).

Reply:
Added